engine/access/rpc/backend/backend.go

```diff
@@ -249,43 +244,104 @@ func convertStorageError(err error) error {
 }
 
 // executionNodesForBlockID returns upto maxExecutionNodesCnt number of randomly chosen execution node identities
-// which have executed the given block ID. If no such execution node is found, an empty list is returned.
+// which have executed the given block ID.
+// If no such execution node is found, an ExecutionReceiptNotFound error is returned.
 func executionNodesForBlockID(
+	ctx context.Context,
 	blockID flow.Identifier,
 	executionReceipts storage.ExecutionReceipts,
 	state protocol.State,
 	log zerolog.Logger) (flow.IdentityList, error) {
 
+	var executorIDs flow.IdentifierList
+	var err error
+	attempt := 0
+	// try to find atleast minExecutionNodesCnt execution node ids from the execution receipts for the given blockID
+	for ; attempt < maxAttemptsForExecutionReceipt; attempt++ {
+
+		executorIDs, err = findAllExecutionNodes(blockID, executionReceipts, log)
+		if err != nil {
+			return flow.IdentityList{}, err
+		}
+
+		if len(executorIDs) >= minExecutionNodesCnt {
+			break
+		}
+
+		// log the attempt
+		log.Debug().Int("attempt", attempt).Int("max_attempt", maxAttemptsForExecutionReceipt).
+			Int("execution_receipts_found", len(executorIDs)).
+			Str("block_id", blockID.String()).
+			Msg("insufficient execution receipts")
+
+		// if one or less execution receipts may have been received then re-query
+		// in the hope that more might have been received by now
+
+		select {
+		case <-ctx.Done():
+			return flow.IdentityList{}, err
+		case <-time.After(100 * time.Millisecond << time.Duration(attempt)):
+			//retry after an exponential backoff
+		}
+	}
+
+	// if no execution receipts have been received so far, then throw an error
+	if len(executorIDs) == 0 {
```

Comment:
return a special error if no execution receipt were found.
We still carry on if only one receipt was found (as before).